File: .a-new-structure/b-word/L0-universal/c-hybrid/foundation/src/session_context.rs
//omni:code --rust -library
//omni:key B-L0-foundation-sessioncontext-rs
//omni:version a-01.00

// ============================================================================
// METADATA
// ============================================================================

//! # Session Context (Rust)
//!
//! Substrate session context schema for runtime integration.
//!
//! This is the Claude Code statusline API schema — generic substrate
//! integration contract, not CPI-SI intelligence.
//!
//! Promoted from L3-cpisi to L0-universal.

// ============================================================================
// SETUP
// ============================================================================

use serde::{Deserialize, Serialize};

// ────────────────────────────────────────────────────────────────
// S.1 Types
// ────────────────────────────────────────────────────────────────

/// Session context from Claude Code hook events.
///
/// The contract between the substrate runtime and our systems.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionContext {
    pub hook_event_name: String,
    pub session_id: String,
    pub transcript_path: String,
    pub cwd: String,
    pub version: String,
    pub model: Model,
    pub workspace: Workspace,
    pub output_style: OutputStyle,
    pub cost: Cost,
    pub context_window: ContextWindow,
}

/// Model identification.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Model {
    pub id: String,
    pub display_name: String,
}

/// Workspace paths.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Workspace {
    pub current_dir: String,
    pub project_dir: String,
}

/// Active output style.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OutputStyle {
    pub name: String,
}

/// Session cost metrics.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Cost {
    pub total_cost_usd: f64,
    pub total_duration_ms: i64,
    pub total_api_duration_ms: i64,
    pub total_lines_added: i64,
    pub total_lines_removed: i64,
}

/// Context window tracking.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContextWindow {
    pub total_input_tokens: i64,
    pub total_output_tokens: i64,
    pub context_window_size: i64,
    pub current_usage: CurrentUsage,
    #[serde(default)]
    pub used_percentage: i64,
    #[serde(default)]
    pub remaining_percentage: i64,
}

/// Current context usage (more accurate than cumulative totals).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CurrentUsage {
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cache_creation_input_tokens: i64,
    pub cache_read_input_tokens: i64,
}

// ============================================================================
// BODY
// ============================================================================

// ────────────────────────────────────────────────────────────────
// B.1 Helper Methods
// ────────────────────────────────────────────────────────────────

impl SessionContext {
    /// Best available working directory: CWD > current_dir > project_dir.
    pub fn work_dir(&self) -> &str {
        if !self.cwd.is_empty() {
            &self.cwd
        } else if !self.workspace.current_dir.is_empty() {
            &self.workspace.current_dir
        } else {
            &self.workspace.project_dir
        }
    }

    /// Total lines changed (added + removed).
    pub fn total_lines_changed(&self) -> i64 {
        self.cost.total_lines_added + self.cost.total_lines_removed
    }

    /// Actual tokens currently in context.
    pub fn current_context_tokens(&self) -> i64 {
        self.context_window.current_usage.input_tokens
            + self.context_window.current_usage.output_tokens
    }

    /// Context usage percentage (0.0 - 100.0).
    pub fn context_percentage(&self) -> f64 {
        if self.context_window.context_window_size == 0 {
            return 0.0;
        }
        let current = self.current_context_tokens();
        if current > 0 {
            return current as f64 / self.context_window.context_window_size as f64 * 100.0;
        }
        let total = self.context_window.total_input_tokens
            + self.context_window.total_output_tokens;
        total as f64 / self.context_window.context_window_size as f64 * 100.0
    }

    /// Cache efficiency: log3(cache/input), capped at +/-3, scaled to -100..+100.
    pub fn cache_efficiency(&self) -> f64 {
        let input = self.context_window.current_usage.input_tokens;
        let cache = self.context_window.current_usage.cache_read_input_tokens;

        if input == 0 && cache == 0 { return 0.0; }
        if input == 0 { return 100.0; }
        if cache == 0 { return -100.0; }

        let ln3 = 3.0_f64.ln();
        let log_ratio = ((cache as f64).ln() - (input as f64).ln()) / ln3;
        let capped = log_ratio.clamp(-3.0, 3.0);
        capped * (100.0 / 3.0)
    }

    /// Session duration in seconds.
    pub fn duration_seconds(&self) -> i64 {
        self.cost.total_duration_ms / 1000
    }

    /// API call duration in seconds.
    pub fn api_duration_seconds(&self) -> i64 {
        self.cost.total_api_duration_ms / 1000
    }

    /// API efficiency: ratio of API time to total time (0.0 - 100.0).
    pub fn api_efficiency(&self) -> f64 {
        if self.cost.total_duration_ms == 0 {
            return 0.0;
        }
        self.cost.total_api_duration_ms as f64 / self.cost.total_duration_ms as f64 * 100.0
    }
}

// ============================================================================
// CLOSING
// ============================================================================
//
// Substrate session context — promoted from L3 to serve all layers.
// "God is not the author of confusion" — 1 Corinthians 14:33

#[cfg(test)]
mod tests {
    use super::*;

    fn test_ctx() -> SessionContext {
        SessionContext {
            hook_event_name: "Status".into(),
            session_id: "test-123".into(),
            transcript_path: "/tmp/test.jsonl".into(),
            cwd: "/home/test".into(),
            version: "1.0.0".into(),
            model: Model { id: "claude-opus-4-6".into(), display_name: "Opus 4.6".into() },
            workspace: Workspace { current_dir: "/home/test/project".into(), project_dir: "/home/test/project".into() },
            output_style: OutputStyle { name: "default".into() },
            cost: Cost { total_cost_usd: 0.5, total_duration_ms: 60000, total_api_duration_ms: 30000, total_lines_added: 100, total_lines_removed: 20 },
            context_window: ContextWindow {
                total_input_tokens: 50000,
                total_output_tokens: 10000,
                context_window_size: 200000,
                current_usage: CurrentUsage { input_tokens: 40000, output_tokens: 8000, cache_creation_input_tokens: 5000, cache_read_input_tokens: 35000 },
                used_percentage: 24,
                remaining_percentage: 76,
            },
        }
    }

    #[test]
    fn work_dir_prefers_cwd() {
        let ctx = test_ctx();
        assert_eq!(ctx.work_dir(), "/home/test");
    }

    #[test]
    fn total_lines() {
        let ctx = test_ctx();
        assert_eq!(ctx.total_lines_changed(), 120);
    }

    #[test]
    fn context_tokens() {
        let ctx = test_ctx();
        assert_eq!(ctx.current_context_tokens(), 48000);
    }

    #[test]
    fn context_pct() {
        let ctx = test_ctx();
        let pct = ctx.context_percentage();
        assert!(pct > 20.0 && pct < 30.0);
    }

    #[test]
    fn api_efficiency() {
        let ctx = test_ctx();
        assert!((ctx.api_efficiency() - 50.0).abs() < 0.01);
    }

    #[test]
    fn cache_efficiency_balanced() {
        // cache_read = 35000, input = 40000 — close to balanced
        let ctx = test_ctx();
        let eff = ctx.cache_efficiency();
        assert!(eff > -20.0 && eff < 0.0); // slightly negative
    }
}

File: .a-new-structure/b-word/L0-universal/c-hybrid/core/foundation/go/types/sessioncontext.go
//omni:code --go -library
//omni:key B-L0-foundation-types-sessioncontext
//omni:version a-02.00

// ============================================================================
// METADATA
// ============================================================================
//
// Key: B-L0-foundation-types-sessioncontext
// Purpose: Substrate session context schema for runtime integration
// Biblical: 1 Corinthians 14:33 - "God is not the author of confusion"
// Authors: Nova Dawn
// Version: 2.0.0
// Created: 2026-02-16
// Updated: 2026-02-18
//
// Promoted from L3-cpisi to L0-universal. This is the Claude Code statusline
// API schema — generic substrate integration, not CPI-SI intelligence.
//
// Full schema from Claude Code statusline API:
//   - Session: hook_event_name, session_id, transcript_path, cwd, version
//   - Model: id, display_name
//   - Workspace: current_dir, project_dir
//   - OutputStyle: name
//   - Cost: total_cost_usd, total_duration_ms, total_api_duration_ms,
//           total_lines_added, total_lines_removed
//   - ContextWindow: total_input_tokens, total_output_tokens,
//                    context_window_size, current_usage
//
// ============================================================================
// END METADATA
// ============================================================================

package types

// ============================================================================
// SETUP
// ============================================================================

import (
	"fmt"
	"math"
)

// --- Core Types ---

// Model represents the substrate model identity.
type Model struct {
	ID          string `json:"id"`
	DisplayName string `json:"display_name"`
}

// Workspace represents spatial grounding — where work happens.
type Workspace struct {
	CurrentDir string `json:"current_dir"`
	ProjectDir string `json:"project_dir"`
}

// OutputStyle represents the identity voice configuration.
type OutputStyle struct {
	Name string `json:"name"`
}

// Cost tracks flow metrics — time, money, lines changed.
type Cost struct {
	TotalCostUSD       float64 `json:"total_cost_usd"`
	TotalDurationMS    int     `json:"total_duration_ms"`
	TotalAPIDurationMS int     `json:"total_api_duration_ms"`
	TotalLinesAdded    int     `json:"total_lines_added"`
	TotalLinesRemoved  int     `json:"total_lines_removed"`
}

// CurrentUsage tracks current API call token usage.
type CurrentUsage struct {
	InputTokens              int `json:"input_tokens"`
	OutputTokens             int `json:"output_tokens"`
	CacheCreationInputTokens int `json:"cache_creation_input_tokens"`
	CacheReadInputTokens     int `json:"cache_read_input_tokens"`
}

// ContextWindow tracks depth — how much context is consumed.
type ContextWindow struct {
	TotalInputTokens    int          `json:"total_input_tokens"`
	TotalOutputTokens   int          `json:"total_output_tokens"`
	ContextWindowSize   int          `json:"context_window_size"`
	CurrentUsage        CurrentUsage `json:"current_usage"`
	UsedPercentage      int          `json:"used_percentage"`
	RemainingPercentage int          `json:"remaining_percentage"`
}

// SessionContext represents the JSON input from Claude Code hooks.
// This is the contract between the substrate runtime and our systems.
type SessionContext struct {
	HookEventName  string        `json:"hook_event_name"`
	SessionID      string        `json:"session_id"`
	TranscriptPath string        `json:"transcript_path"`
	CWD            string        `json:"cwd"`
	Version        string        `json:"version"`
	Model          Model         `json:"model"`
	Workspace      Workspace     `json:"workspace"`
	OutputStyle    OutputStyle   `json:"output_style"`
	Cost           Cost          `json:"cost"`
	ContextWindow  ContextWindow `json:"context_window"`
}

// ============================================================================
// END SETUP
// ============================================================================

// ============================================================================
// BODY
// ============================================================================

// ──────────────────────────────────────────────────────────────────────────
// 1. Org Chart
// ──────────────────────────────────────────────────────────────────────────
//
// Queries (SessionContext):
//   GetWorkDir, TotalLinesChanged, CurrentContextTokens, ContextPercentage,
//   CacheEfficiency, DurationSeconds, APIDurationSeconds, APIEfficiency, String
//
// Flow: JSON unmarshal → SessionContext → query methods → formatted output

// ──────────────────────────────────────────────────────────────────────────
// 2. Helpers
// ──────────────────────────────────────────────────────────────────────────
// [Reserved: No internal helpers — uses only math.Log from stdlib]

// ──────────────────────────────────────────────────────────────────────────
// 3. Core Operations
// ──────────────────────────────────────────────────────────────────────────

// --- Workspace Queries ---

// GetWorkDir returns the best available working directory.
// Priority: CWD > CurrentDir > ProjectDir
func (ctx *SessionContext) GetWorkDir() string {
	if ctx.CWD != "" {
		return ctx.CWD
	}
	if ctx.Workspace.CurrentDir != "" {
		return ctx.Workspace.CurrentDir
	}
	return ctx.Workspace.ProjectDir
}

// --- Cost Queries ---

// TotalLinesChanged returns sum of added and removed lines.
func (ctx *SessionContext) TotalLinesChanged() int {
	return ctx.Cost.TotalLinesAdded + ctx.Cost.TotalLinesRemoved
}

// DurationSeconds returns session duration in seconds.
func (ctx *SessionContext) DurationSeconds() int {
	return ctx.Cost.TotalDurationMS / 1000
}

// APIDurationSeconds returns API call duration in seconds.
func (ctx *SessionContext) APIDurationSeconds() int {
	return ctx.Cost.TotalAPIDurationMS / 1000
}

// APIEfficiency returns ratio of API time to total time (0-100).
// Lower = more thinking time, higher = more API waiting.
func (ctx *SessionContext) APIEfficiency() float64 {
	if ctx.Cost.TotalDurationMS == 0 {
		return 0
	}
	return float64(ctx.Cost.TotalAPIDurationMS) / float64(ctx.Cost.TotalDurationMS) * 100
}

// --- Context Window Queries ---

// CurrentContextTokens returns actual tokens in context (from current_usage).
// More accurate than TotalInputTokens which is cumulative.
func (ctx *SessionContext) CurrentContextTokens() int {
	return ctx.ContextWindow.CurrentUsage.InputTokens +
		ctx.ContextWindow.CurrentUsage.OutputTokens
}

// ContextPercentage returns current context usage as percentage (0-100).
//
// Uses CurrentUsage for accuracy. Falls back to cumulative totals
// when CurrentUsage is unavailable (zero).
func (ctx *SessionContext) ContextPercentage() float64 {
	if ctx.ContextWindow.ContextWindowSize == 0 {
		return 0
	}

	// Prefer current usage (more accurate)
	tokens := ctx.CurrentContextTokens()
	if tokens == 0 {
		// Fallback to cumulative
		tokens = ctx.ContextWindow.TotalInputTokens + ctx.ContextWindow.TotalOutputTokens
	}

	return float64(tokens) / float64(ctx.ContextWindow.ContextWindowSize) * 100
}

// CacheEfficiency returns cache efficiency using base-3 log-odds scaling.
//
// Scale: log3(cache/input), capped at +/-3 trits, mapped to -100..+100.
//
//	Scale: -100 (all new input) → ±0 (balanced) → +100 (all cache)
//
// Edge cases:
//   - Both zero: 0 (balanced — no data)
//   - Input zero, cache > 0: +100 (all cache)
//   - Cache zero, input > 0: -100 (all new)
//   - Equal: 0 (balanced)
//
// "A just weight and balance are the LORD's" — Proverbs 16:11
func (ctx *SessionContext) CacheEfficiency() float64 {
	input := ctx.ContextWindow.CurrentUsage.InputTokens
	cacheRead := ctx.ContextWindow.CurrentUsage.CacheReadInputTokens

	if input == 0 && cacheRead == 0 {
		return 0
	}
	if input == 0 {
		return 100
	}
	if cacheRead == 0 {
		return -100
	}

	ln3 := math.Log(3)
	logRatio := (math.Log(float64(cacheRead)) - math.Log(float64(input))) / ln3

	// Clamp to [-3, +3] trits
	if logRatio > 3 {
		logRatio = 3
	} else if logRatio < -3 {
		logRatio = -3
	}

	return logRatio * (100.0 / 3.0)
}

// ──────────────────────────────────────────────────────────────────────────
// 4. Error Handling
// ──────────────────────────────────────────────────────────────────────────
// [Reserved: Types package — query methods return values, not errors]

// ──────────────────────────────────────────────────────────────────────────
// 5. Public APIs
// ──────────────────────────────────────────────────────────────────────────

// --- String Methods ---

// String returns the model display name and ID.
//
//	"Claude Opus 4.6 (claude-opus-4-6)"
func (m Model) String() string {
	if m.DisplayName == "" && m.ID == "" {
		return "(unknown model)"
	}
	if m.DisplayName == "" {
		return m.ID
	}
	return fmt.Sprintf("%s (%s)", m.DisplayName, m.ID)
}

// String returns a human-readable summary of the session context.
//
//	"[session-abc] Claude Opus 4.6 (claude-opus-4-6) — Notification (42.5% context, $0.12)"
func (ctx *SessionContext) String() string {
	pct := ctx.ContextPercentage()
	return fmt.Sprintf("[%s] %s — %s (%.1f%% context, $%.2f)",
		ctx.SessionID,
		ctx.Model.String(),
		ctx.HookEventName,
		pct,
		ctx.Cost.TotalCostUSD,
	)
}

// ============================================================================
// END BODY
// ============================================================================

// ============================================================================
// CLOSING
// ============================================================================
//
// ──────────────────────────────────────────────────────────────────────────
// Cv — Validation
// ──────────────────────────────────────────────────────────────────────────
//
// Testing: go test -v -run TestSessionContext ./L0-universal/.../types/
//
// Verify:
//   - JSON roundtrip with named types matches original inline layout
//   - ContextPercentage falls back to cumulative when CurrentUsage is zero
//   - CacheEfficiency edge cases: both zero, one zero, equal, extreme ratios
//   - GetWorkDir cascade: CWD > CurrentDir > ProjectDir
//
// ──────────────────────────────────────────────────────────────────────────
// Ce — Execution
// ──────────────────────────────────────────────────────────────────────────
//
// Library — imported, not executed.
// import "creativeworkzstudio.com/bereshit/L0-universal/c-hybrid/core/foundation/go/types"
//
// ──────────────────────────────────────────────────────────────────────────
// Cc — Cleanup
// ──────────────────────────────────────────────────────────────────────────
//
// No resources to clean up. Pure value types.
//
// ──────────────────────────────────────────────────────────────────────────
// X1: Policy
// ──────────────────────────────────────────────────────────────────────────
//
// Never:
//   - Change JSON field names — breaks Claude Code hook contract
//   - Add filesystem I/O — types only
//   - Change CacheEfficiency scaling — breaks health scoring
//
// Careful:
//   - Adding new fields to SessionContext — verify against Claude Code API
//   - Modifying ContextPercentage fallback — affects context warnings
//
// Safe:
//   - Adding new query methods
//   - Adding constructors for sub-types
//
// ──────────────────────────────────────────────────────────────────────────
// X5: Note
// ──────────────────────────────────────────────────────────────────────────
//
// Note: Substrate session context promoted from L3 to L0. Named types
// extracted from inline definitions for testability and reuse. JSON layout
// unchanged. Identity vars (Pragma/Metadata) live in workflow.go — Go
// requires unique package-level names across files in the same package.
//
// Scripture: 1 Corinthians 14:33 — "God is not the author of confusion,
// but of peace."
//
// ============================================================================
// END CLOSING
// ============================================================================
